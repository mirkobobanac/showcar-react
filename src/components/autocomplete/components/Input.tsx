import * as React from 'react'

type IInputProps = {
  id?: string
  placeholder: string
  value: string | undefined
  error: boolean | undefined
  disabled?: boolean
  hideArrow?: boolean
  onChange: (search: string) => void
  onKeyUpArrow: (e: React.KeyboardEvent<HTMLInputElement>) => void
  onKeyDownArrow: (e: React.KeyboardEvent<HTMLInputElement>) => void
  onKeyEscape: (e: React.KeyboardEvent<HTMLInputElement>) => void
  onKeyEnter: (e: React.KeyboardEvent<HTMLInputElement>) => void
  onClick: (e: React.MouseEvent<HTMLInputElement>) => void
  onDropdownClick: (inputFocus: () => void) => void
  /**
   * Triggered when the input is reset
   */
  onClearClick: () => void
  blurOnKeyEscape: boolean
}

const Input: React.SFC<IInputProps> = props => {
  let inputRef: HTMLInputElement | null

  const focus = () => inputRef && inputRef.focus()
  const blur = () => inputRef && inputRef.blur()

  return (
    <div className="react-autocomplete__input-wrapper">
      <input
        id={props.id || undefined}
        type="text"
        data-role="user-query"
        className={`react-autocomplete__input ${props.error ? 'error' : ''}`}
        data-ignore="true"
        placeholder={props.placeholder}
        value={props.value}
        ref={input => {
          inputRef = input
        }}
        onKeyDown={(e: React.KeyboardEvent<HTMLInputElement>) => {
          // https://stackoverflow.com/questions/5597060/detecting-arrow-key-presses-in-javascript
          switch (e.keyCode) {
            case 27:
              props.onKeyEscape(e)
              if (props.blurOnKeyEscape) {
                blur()
              }
              return
            case 13:
              props.onKeyEnter(e)
              return blur()
            case 38:
              e.preventDefault() // we don't to move input cursor on keypress
              return props.onKeyUpArrow(e)
            case 40:
              e.preventDefault() // we don't to move input cursor on keypress
              return props.onKeyDownArrow(e)
            default:
          }
        }}
        onChange={(e: React.ChangeEvent<HTMLInputElement>) => props.onChange(e.target.value)}
        onClick={props.onClick}
        disabled={props.disabled}
      />

      <div className="react-autocomplete__icon-wrapper">
        {!(props.error || (props.value && props.value.length > 0)) ? (
          <div className="react-autocomplete__icon-dropdown" onClick={() => props.onDropdownClick(focus)}>
            {props.hideArrow !== true && <as24-icon class="react-autocomplete__icon-dropdown__icon" type="arrow" />}
          </div>
        ) : (
          <div
            className="react-autocomplete__icon-cross"
            onClick={() => {
              props.onChange('') // clear input
              focus()
              props.onClearClick()
            }}
          >
            <as24-icon class="react-autocomplete__icon-dropdown__icon" type="close" />
          </div>
        )}
      </div>
    </div>
  )
}
export default Input
