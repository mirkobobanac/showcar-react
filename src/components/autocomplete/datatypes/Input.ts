import Record, { RecordR } from '../../../immutability/Record'

export type IImmutableInput<T> = RecordR<IInput<T>>

export type IInput<T> = {
  id: T
  label: string
}

export const immutableInput = <T>(item: IInput<T>): IImmutableInput<T> => Record.new(item)

export const itemMatchesSearch = <T>(search: string) => (item: IImmutableInput<T>): boolean =>
  item.label.toLowerCase().indexOf(search.toLowerCase()) !== -1
