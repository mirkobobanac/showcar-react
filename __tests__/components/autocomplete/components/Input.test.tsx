import Enzyme, { mount, render, shallow } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
import 'jest'
import * as React from 'react'
import Input from '../../../../src/components/autocomplete/components/Input'

Enzyme.configure({ adapter: new Adapter() })

const inputWithArrowIcon = (
  <Input
    placeholder={'Alle'}
    value={''}
    error={false}
    onChange={() => {}}
    onKeyUpArrow={() => {}}
    onKeyDownArrow={() => {}}
    onKeyEscape={() => {}}
    onKeyEnter={() => {}}
    onClick={() => {}}
    onDropdownClick={() => {}}
    onClearClick={() => {}}
    blurOnKeyEscape={true}
  />
)

const inputWithCloseIcon = (
  <Input
    placeholder={'Alle'}
    value={'abc'}
    error={false}
    onChange={() => {}}
    onKeyUpArrow={() => {}}
    onKeyDownArrow={() => {}}
    onKeyEscape={() => {}}
    onKeyEnter={() => {}}
    onClick={() => {}}
    onDropdownClick={() => {}}
    onClearClick={() => {}}
    blurOnKeyEscape={true}
  />
)

const inputWithoutIcon = (
  <Input
    placeholder={'Alle'}
    value={''}
    error={false}
    onChange={() => {}}
    onKeyUpArrow={() => {}}
    onKeyDownArrow={() => {}}
    onKeyEscape={() => {}}
    onKeyEnter={() => {}}
    onClick={() => {}}
    onDropdownClick={() => {}}
    onClearClick={() => {}}
    blurOnKeyEscape={true}
    hideArrow={true}
  />
)

test('Input is rendered with arrow icon', () => {
  const input = render(inputWithArrowIcon)

  expect(input.find('as24-icon').attr('type')).toEqual('arrow')

  expect(input).toMatchSnapshot()
})

test('Input is rendered with close icon', () => {
  const input = render(inputWithCloseIcon)

  expect(input.find('as24-icon').attr('type')).toEqual('close')
})

test('Input is rendered without an icon', () => {
  const input = render(inputWithoutIcon)

  expect(input.find('as24-icon').get().length).toEqual(0)
})
