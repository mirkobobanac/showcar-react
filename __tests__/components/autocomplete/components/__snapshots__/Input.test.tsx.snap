// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input is rendered with arrow icon 1`] = `
initialize {
  "0": Object {
    "attribs": Object {
      "class": "react-autocomplete__input-wrapper",
    },
    "children": Array [
      Object {
        "attribs": Object {
          "class": "react-autocomplete__input ",
          "data-ignore": "true",
          "data-role": "user-query",
          "placeholder": "Alle",
          "type": "text",
          "value": "",
        },
        "children": Array [],
        "name": "input",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": Object {
          "attribs": Object {
            "class": "react-autocomplete__icon-wrapper",
          },
          "children": Array [
            Object {
              "attribs": Object {
                "class": "react-autocomplete__icon-dropdown",
              },
              "children": Array [
                Object {
                  "attribs": Object {
                    "class": "react-autocomplete__icon-dropdown__icon",
                    "type": "arrow",
                  },
                  "children": Array [],
                  "name": "as24-icon",
                  "namespace": "http://www.w3.org/1999/xhtml",
                  "next": null,
                  "parent": [Circular],
                  "prev": null,
                  "type": "tag",
                  "x-attribsNamespace": Object {
                    "class": undefined,
                    "type": undefined,
                  },
                  "x-attribsPrefix": Object {
                    "class": undefined,
                    "type": undefined,
                  },
                },
              ],
              "name": "div",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": null,
              "parent": [Circular],
              "prev": null,
              "type": "tag",
              "x-attribsNamespace": Object {
                "class": undefined,
              },
              "x-attribsPrefix": Object {
                "class": undefined,
              },
            },
          ],
          "name": "div",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": null,
          "parent": [Circular],
          "prev": [Circular],
          "type": "tag",
          "x-attribsNamespace": Object {
            "class": undefined,
          },
          "x-attribsPrefix": Object {
            "class": undefined,
          },
        },
        "parent": [Circular],
        "prev": null,
        "type": "tag",
        "x-attribsNamespace": Object {
          "class": undefined,
          "data-ignore": undefined,
          "data-role": undefined,
          "placeholder": undefined,
          "type": undefined,
          "value": undefined,
        },
        "x-attribsPrefix": Object {
          "class": undefined,
          "data-ignore": undefined,
          "data-role": undefined,
          "placeholder": undefined,
          "type": undefined,
          "value": undefined,
        },
      },
      Object {
        "attribs": Object {
          "class": "react-autocomplete__icon-wrapper",
        },
        "children": Array [
          Object {
            "attribs": Object {
              "class": "react-autocomplete__icon-dropdown",
            },
            "children": Array [
              Object {
                "attribs": Object {
                  "class": "react-autocomplete__icon-dropdown__icon",
                  "type": "arrow",
                },
                "children": Array [],
                "name": "as24-icon",
                "namespace": "http://www.w3.org/1999/xhtml",
                "next": null,
                "parent": [Circular],
                "prev": null,
                "type": "tag",
                "x-attribsNamespace": Object {
                  "class": undefined,
                  "type": undefined,
                },
                "x-attribsPrefix": Object {
                  "class": undefined,
                  "type": undefined,
                },
              },
            ],
            "name": "div",
            "namespace": "http://www.w3.org/1999/xhtml",
            "next": null,
            "parent": [Circular],
            "prev": null,
            "type": "tag",
            "x-attribsNamespace": Object {
              "class": undefined,
            },
            "x-attribsPrefix": Object {
              "class": undefined,
            },
          },
        ],
        "name": "div",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": null,
        "parent": [Circular],
        "prev": Object {
          "attribs": Object {
            "class": "react-autocomplete__input ",
            "data-ignore": "true",
            "data-role": "user-query",
            "placeholder": "Alle",
            "type": "text",
            "value": "",
          },
          "children": Array [],
          "name": "input",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": [Circular],
          "parent": [Circular],
          "prev": null,
          "type": "tag",
          "x-attribsNamespace": Object {
            "class": undefined,
            "data-ignore": undefined,
            "data-role": undefined,
            "placeholder": undefined,
            "type": undefined,
            "value": undefined,
          },
          "x-attribsPrefix": Object {
            "class": undefined,
            "data-ignore": undefined,
            "data-role": undefined,
            "placeholder": undefined,
            "type": undefined,
            "value": undefined,
          },
        },
        "type": "tag",
        "x-attribsNamespace": Object {
          "class": undefined,
        },
        "x-attribsPrefix": Object {
          "class": undefined,
        },
      },
    ],
    "name": "div",
    "namespace": "http://www.w3.org/1999/xhtml",
    "next": null,
    "parent": null,
    "prev": null,
    "root": Object {
      "attribs": Object {},
      "children": Array [
        [Circular],
      ],
      "name": "root",
      "namespace": "http://www.w3.org/1999/xhtml",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
      "x-attribsNamespace": Object {},
      "x-attribsPrefix": Object {},
    },
    "type": "tag",
    "x-attribsNamespace": Object {
      "class": undefined,
    },
    "x-attribsPrefix": Object {
      "class": undefined,
    },
  },
  "_root": [Circular],
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "normalizeWhitespace": false,
    "withDomLvl1": true,
    "xml": false,
  },
}
`;
